services/command-router-base/src/main/java/org/eclipse/hono/commandrouter/impl/KubernetesBasedAdapterInstanceStatusService.java

```diff
@@ -0,0 +1,344 @@
+                        podName, shortContainerId, containerStatus.getState().getTerminated().getReason(), containerIdToPodNameMap.size());
+                terminatedContainerIds.add(shortContainerId);
+                onAdapterContainerRemoved(shortContainerId);
+            }
+        } else {
+            if (watchAction == Watcher.Action.MODIFIED) {
+                final String oldContainerId = podNameToContainerIdMap.get(podName);
+                // perform check to make sure that we didn't miss a "terminated container" event
+                if (oldContainerId != null && !oldContainerId.equals(shortContainerId)) {
+                    podNameToContainerIdMap.remove(podName, oldContainerId);
+                    if (containerIdToPodNameMap.remove(oldContainerId) != null) {
+                        LOG.info("removed obsolete entry for pod [{}], container [{}]; active adapter containers now: {}",
+                                podName, oldContainerId, containerIdToPodNameMap.size());
+                        terminatedContainerIds.add(oldContainerId);
+                        onAdapterContainerRemoved(oldContainerId);
+                    }
+                }
+            }
+            podNameToContainerIdMap.put(podName, shortContainerId);
+            if (containerIdToPodNameMap.put(shortContainerId, podName) == null) {
+                LOG.info("added entry for pod [{}], container [{}]; active adapter containers now: {}", podName,
+                        shortContainerId, containerIdToPodNameMap.size());
+                suspectedContainerIds.remove(shortContainerId);
+                onAdapterContainerAdded(shortContainerId);
+            }
+        }
+    }
+
+    /**
+     * Invoked when an adapter container was added.
+     * <p>
+     * This method does nothing by default and may be overridden by subclasses.
+     *
+     * @param containerId The container identifier.
+     */
+    protected void onAdapterContainerAdded(final String containerId) {
+       // nothing done by default
+    }
+
+    /**
+     * Invoked when an adapter container was deleted or terminated.
+     * <p>
+     * This method does nothing by default and may be overridden by subclasses.
+     *
+     * @param containerId The container identifier.
+     */
+    protected void onAdapterContainerRemoved(final String containerId) {
+        // nothing done by default
+    }
+
+    private void onWatcherClosed(final WatcherException cause) {
+        if (active.compareAndExchange(1, 0) != 1) {
+            return; // watcher already closed or status service already stopped
+        }
+        containerIdToPodNameMap.clear();
+        podNameToContainerIdMap.clear();
+        LOG.error("Watcher closed with error", cause);
+        while (active.get() == 0) {
+            try {
+                Thread.sleep(WATCH_RECREATION_DELAY_MILLIS);
```

Comment:
this looks very, very strange ...

Reply:
I see. Does the kubernetes client say anything about the impact of blocking the thread that this method is invoked on? You might end up blocking the thread for quite some time ...

Reply:
Looking at KubernetesClient / OkHttp DEBUG log output when running the `testServiceReactivatesAfterWatchClosed` test method, I don't see an issue here.
Testing with a sleep period of 60s, I still see maintenance work on the connection getting done in this period via other threads:
```
09:19:36.444 [OkHttp TaskRunner] DEBUG o.internal.concurrent.TaskRunner - Q10003 starting              : OkHttp WebSocket http://localhost:42369/... ping
09:19:36.446 [OkHttp WebSocket http://localhost:42369/... ping] DEBUG o.internal.concurrent.TaskRunner - Q10003 run again after  30 s : OkHttp WebSocket http://localhost:42369/... ping
09:19:36.447 [OkHttp TaskRunner] DEBUG o.internal.concurrent.TaskRunner - Q10003 finished run in   2 ms: OkHttp WebSocket http://localhost:42369/... ping
```
And in terms of the KubernetesClient, the watcher is closed at that point, so nothing to be done there.

Reply:
ok